src/Microsoft.AspNet.Mvc.Core/Formatters/MediaType.cs

```
@@ -0,0 +1,678 @@
+
+            _parameterEnumerator = enumerator;
+        }
+
+        /// <summary>
+        /// Gets the type of the <see cref="MediaType"/>.
+        /// </summary>
+        public StringSegment Type { get; }
+
+        /// <summary>
+        /// Gets whether this <see cref="MediaType"/> matches all types.
+        /// </summary>
+        public bool MatchesAllTypes => Type.Equals("*", StringComparison.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Gets the subtype of the <see cref="MediaType"/>.
+        /// </summary>
+        public StringSegment SubType { get; private set; }
+
+        /// <summary>
+        /// Gets whether this <see cref="MediaType"/> matches all subtypes.
+        /// </summary>
+        public bool MatchesAllSubTypes => SubType.Equals("*", StringComparison.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Gets whether this <see cref="MediaType"/> has a valid value.
+        /// </summary>
+        public bool HasValue => !_parameterEnumerator.ParsingFailed;
+
+        /// <summary>
+        /// Gets the <see cref="System.Text.Encoding"/> of the <see cref="MediaType"/> if it has one.
+        /// </summary>
+        public Encoding Encoding => GetEncodingFromCharset(GetParameter("charset"));
+
+        /// <summary>
+        /// Gets the charset parameter of the <see cref="MediaType"/> if it has one.
+        /// </summary>
+        public StringSegment Charset => GetParameter("charset");
+
+        /// <summary>
+        /// Determines whether the current <see cref="MediaType"/> is a subset of the <paramref name="set"/> <see cref="MediaType"/>.
+        /// </summary>
+        /// <param name="set">The set <see cref="MediaType"/>.</param>
+        /// <returns>
+        /// <code>true</code> if this <see cref="MediaType"/> is a subset of <paramref name="set"/>; otherwise<code>false</code>.
+        /// </returns>
+        public bool IsSubsetOf(MediaType set)
+        {
+            return HasValue && set.HasValue &&
```

> I think both things are correct, what advantage does throwing here provide?

Throwing lets you know if you're in an unexpected case. Generally if you electrify the fences you need to take steps to avoid touching them. This can either be good or bad. Generally throwing is the safer choice if you don't understand 100% of the scenarios.


> IMO it's ok to say that if you don't give me two valid MediaTypes IsSubsetOf is going to be false all the time, in the current usage on our code, it would imply that a formatter doesn't get selected which I think it's the right behavior.

I'm not 100% convinced about this. What you're describing is a silent failure.